src/main/java/net/jodah/typetools/TypeResolver.java

```diff
@@ -334,6 +408,60 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
     return map;
   }
 
+  private static Type substituteGenerics(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap) {
+    if (genericType == null) {
+      return null;
+    }
+    if (genericType instanceof Class<?>) {
+      return genericType;
+    }
+    if (genericType instanceof TypeVariable<?>) {
+      final TypeVariable<?> typeVariable = (TypeVariable<?>) genericType;
+      final Type mapping = typeVariableMap.get(typeVariable);
+      if (mapping != null) {
+        return substituteGenerics(mapping, typeVariableMap);
+      }
+      final Type[] upperBounds = typeVariable.getBounds();
+      // NOTE: According to https://docs.oracle.com/javase/tutorial/java/generics/bounded.html
+      // if there are multiple upper bounds where one bound is a class, then this must be the
+      // leftmost/first bound. Therefore we blindly take this one, hoping is the most relevant.
+      return substituteGenerics(upperBounds[0], typeVariableMap);
+    }
+    if (genericType instanceof WildcardType) {
+      final WildcardType wildcardType = (WildcardType) genericType;
+      final Type[] upperBounds = wildcardType.getUpperBounds();
+      final Type[] lowerBounds = wildcardType.getLowerBounds();
+      if (upperBounds.length == 1 && lowerBounds.length == 0) {
+        return substituteGenerics(upperBounds[0], typeVariableMap);
+      }
+      throw new UnsupportedOperationException(
+        "Attempted to resolve type " + wildcardType + " which has " + upperBounds.length + " upper bounds " +
+        "and " + lowerBounds.length + " lower bounds. " +
+        "Resolution of wildcard types is only supported for the trivial case of exactly one upper bound " +
+        "and no lower bounds. If you require resolution in this more complex case, please file an issue via " +
+        NEW_ISSUE_URL
+      );
+    }
+    if (genericType instanceof ParameterizedType) {
+      final ParameterizedType parameterizedType = (ParameterizedType) genericType;
+      final Type[] actualTypeArguments =  parameterizedType.getActualTypeArguments();
+      final Type[] resolvedTypeArguments = new Type[actualTypeArguments.length];
+
+      boolean changed = false;
+      for (int i = 0; i < actualTypeArguments.length; i++) {
+        resolvedTypeArguments[i] = substituteGenerics(actualTypeArguments[i], typeVariableMap);
+        changed = changed || (resolvedTypeArguments[i] != actualTypeArguments[i]);
+      }
+
+      return changed ? new ResolvedParameterizedType(parameterizedType,  resolvedTypeArguments) : parameterizedType;
+    }
+    throw new UnsupportedOperationException(
```

Comment:
Same re: the unsupported message.

Reply:
Resolved in fc7cffbf0ae21d21b609c6e407458b3e16d5b474